keps/sig-api-machinery/20190723-coordinated-discovery-document.md

```diff
@@ -0,0 +1,181 @@
+   * [Calculating DiscoveryDocHashes](#calculating-discoverydochashes)
+   * [Periodically reporting DiscoveryDocHashes &amp; Garbage Collection](#periodically-reporting-discoverydochashes--garbage-collection)
+   * [Handling discovery requests](#handling-discovery-requests)
+* [Backwards Compatibility](#backwards-compatibility)
+* [Graduation Plan](#graduation-plan)
+* [Alternatives](#alternatives)
+
+## Problem Statement
+
+API servers today compute discovery documents based on their local states. When
+HA API servers are rolling upgraded, API server instances can return
+different discovery documents to clients. This causes problems. For
+example, the storage migrator does not work reliably with HA API servers
+([details][]).
+
+[details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
+
+This KEP proposes a mechanism for HA API servers to expose to clients if
+all API server instances are serving consistent discovery document.
+
+## Summary of Proposed changes
+
+Periodically, each API server instance in an HA setup calculates a hash based on
+the discovery documents of the built-in APIs. The hash value is reported in a
+new API, `DiscoveryDocHashes`, together with the API server's ID. 
+
+When responding to discovery document requests, apart from the original
+discovery documents, the API server also sets a `Consistent` field to indicate if
+all API servers are serving the same discovery documents, i.e., if all hash
+values recorded in `DiscoveryDocHashes` are equal. 
+
+## API changes
+
+We introduce a new API `DiscoveryDocHashes`, in a new API group
+`APIServerCoordination.k8s.io`. 
+
+```golang
+type DiscoveryDocHashes struct {
+   Hashes []Hash
+}
+
+type Hash struct {
+   // The ID of the reporting API server. 
+   // The ID can be generated in a similar fashion as the controller manager
+   // generates leader election id at
+   // https://github.com/kubernetes/kubernetes/blob/2321d1e9e8950cc94b3ef2368dfaacec61f1ba4f/cmd/cloud-controller-manager/app/controllermanager.go#L185.
+   // Required.
+   APIServerID string
+
+   // A hash calculated based on the discovery documents of built-in APIs. If
+   // two hashes equal, it means that the two API servers serve the same set of
+   // APIs.
+   // Required.
+   BuiltInAPIHash string
+
+   // LastHeartbeatTime is the last time the reporting API server updates this 
+   // field.
+   // Required 
+   LastHeartbeatTime metav1.Time
+}
+```
+
+API server instances report the hashes of their discovery documents through this
+API.  Although the main users of this API are the API servers, we make this API
+public to make debugging convenient.
+
+Another API change is that at each level of the discovery [API][] (APIGroupList,
+APIVersions, and APIResourceList), we add a new field `Consistent` to indicate
+if all API servers agree on the discovery documents, i.e., if all hashes
+reported in `DiscoveryDocHashes` are equal.
```

Comment:
I think rather there should be a proposed / accepted categorization, and apiservers always serve the most recent version that is accepted by all apiservers. Don't force clients to understand our coordination system.